### PR TITLE
Enforce conversion method return type & redefine Comparable

### LINKED ISSUE #10355: Comparable.from doesn't return Comparable
- #8502 implemented a way to check return type of a function. 

Alas, conversion methods are not checked right now, because [Comparable.from](https://github.com/enso-org/enso/blob/8fba4b7f23dcb248dcf406a4d9889f93b54998b5/docs/syntax/functions.md?plain=1#L334) typically violates the required contract - it doesn't return `Comparable`, but some unrelated type.

The goal of this issue is to change the way we deal with `Comparable` and turn on the return type check for conversion methods.

```ruby
type Comparable
    private Both value:Any comparator:Any
    
    new value:Any comparator:Any -> Comparable = Comparable.Both value comparator

    // implement operations by using self.value and that.value and delegating to self.comparator
    <,>,>=,<=,=
```

- Verify the extra abstraction doesn't influence benchmarks.
- Turn return type check on: [patch is here](https://github.com/enso-org/enso/issues/10355#issuecomment-2188413299)

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Fixes #10355 by enforcing return type of `Xyz.from` conversion methods. Most frequent violations are caused by _ordering support_ - hence 280b7a7357d07b87998f5e0198c990024d9682d7 defines `Comparable.new` to address the problem:
- use `Ordering.compare` to compare two values to be _equal, less or greater_ or uncomparable
- use `Ordering.hash` to compute hash code for any object in Enso
- use `Comparable.new` while implementing `Comparable.from` conversion to control `compare` and `hash` for an object

Those are the most important API changes. Other types, like `Default_Comparator` were moved outside of the API into `private` module.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary - **done** in ecfcecee48b95c516f1014377ba3b4c642a48307
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides. 
- [x] Unit tests have been written where possible.
- [x] Benchmarks continue to be fast

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
With https://github.com/enso-org/enso/commit/280b7a7357d07b87998f5e0198c990024d9682d7 we seem to get following results:
```
sbt:enso> runEngineDistribution --run test/Base_Tests
2450 tests succeeded.
401 tests failed.
38 tests skipped.
9 groups skipped.
```
Fixing:
- down to **178 tests failed** with e908a32
- down to **134 tests failed** after 5ed87b2
- at **112 failing tests** after 4443559

**Comment by jdunkerley** [comment]
Can we still return errors from
conversions as we use that?

**Comment by JaroslavTulach** [comment]
> Can we still return errors from conversions as we use that?

Yes. `Error.throw` - e.g. `DataflowErrors` are [special "all fit" values](https://github.com/enso-org/enso/blob/295260393187063ebbd74bbdf8d78066a9328842/engine/runtime/src/main/java/org/enso/interpreter/node/callable/argument/ReadArgumentCheckNode.java#L176) and they pass thru without any check.

**Review by jdunkerley — APPROVED** [review]

**Review by GregoryTravis — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
> at **112 failing tests** after [4443559](https://github.com/enso-org/enso/commit/44435591210684a6f394d5ffb549ab2f2776a973)

At **33  failing tests** with 18dde9b as this [report confirms](https://github.com/enso-org/enso/runs/27177820206#r1). Time to _execute some benchmarks_:
- [engine](https://github.com/enso-org/enso/actions/runs/9844138090) - **good**!
- [stdlibs](https://github.com/enso-org/enso/actions/runs/9844142054) - looking **good** as well

**Comment by JaroslavTulach** [comment]
The problem for today is that following **program works**:
```ruby
from Standard.Base import Vector, Comparable

type T
    Value a b

type T_Comparator
    compare t1 t2 = Comparable.from t1.a . compare t2.a
    hash t = Comparable.from t.a . hash

Comparable.from (that:T) = Comparable.new that T_Comparator

main = [T.Value 1 2, T.Value 1 2] . distinct
```
but when one changes the first line to
```ruby
from Standard.Base import all
```
then **it doesn't** and yields:
```
Execution finished with an error: Assertion Error: 'Expecting Ordering or Nothing, but got: Error:Error (By 1 Decimal_Comparator) (By 1 Default_Comparator) with type Error calling T_Comparator.compare[v.enso:7:5-55] self=_ t1=_ t2=_'
        at <enso> Map.get(Map.enso:220:40-70)
```
**Update**: workarounded by a639364ac1

**Comment by JaroslavTulach** [comment]
Time to re-run benchmarks:
- [engine](https://github.com/enso-org/enso/actions/runs/9861151237)
- [stdlib](https://github.com/enso-org/enso/actions/runs/9862022743) with b652893 fix

Engine benchmarks are looking good. StdLibs benchmarks look fine, except:

Table_Sorting_table_sort_objects

Vector_Operations_Max_Stats

Vector_Sorting_vector_sort_objects

which demonstrate something has been changed.

**Comment by JaroslavTulach** [comment]
There is only a [single remaining failure](https://github.com/enso-org/enso/actions/runs/9869132872/job/27252332462?pr=10468#step:7:2416):
```
 ️ [FAILED] [PostgreSQL]  Interactions Between various operations: [9/10, 924ms]
 ️ 
 ️     - [FAILED] aggregates and distinct [63ms]
 ️         Reason: An unexpected dataflow error ((Incomparable_Values.Error Nothing Nothing)) has been matched (at test/Table_Tests/src/Common_Table_Operations/Integration_Tests.enso:70:13-72).
 ️         at <enso> Incomparable_Values.type.handle_errors.handle.case handle(built-distribution/enso-engine-2024.3.1-dev-linux-amd64/enso-2024.3.1-dev/lib/Standard/Base/2024.3.1-dev/src/Errors/Common.enso:278:47-101)
 ️         at <enso> Panic.catch(Internal)
 ️         at <enso> Incomparable_Values.type.handle_errors<arg-1>(built-distribution/enso-engine-2024.3.1-dev-linux-amd64/enso-2024.3.1-dev/lib/Standard/Base/2024.3.1-dev/src/Errors/Common.enso:281-282)
 ️         at <enso> Panic.catch(Internal)
 ️         at <enso> Incomparable_Values.type.handle_errors(built-distribution/enso-engine-2024.3.1-dev-linux-amd64/enso-2024.3.1-dev/lib/Standard/Base/2024.3.1-dev/src/Errors/Common.enso:281-282)
 ️         at <enso> Table.sort<arg-1>(built-distribution/enso-engine-2024.3.1-dev-linux-amd64/enso-2024.3.1-dev/lib/Standard/Table/2024.3.1-dev/src/Table.enso:916-917)
```
how do I **run the PostgreSQL** tests? Usage of `docker` is [described here](https://github.com/enso-org/enso/blob/ee0175c3644e0d3ab1165ee0e1b6f94cc90c8a2b/test/Table_Tests/src/Database/README.md#setup-via-docker). I can start the `docker` and then as #10501 suggests:
```
enso$ ENSO_POSTGRES_DATABASE=postgres \
  ENSO_POSTGRES_PASSWORD=pwd \
  ENSO_POSTGRES_USER=postgres \
  ./built-distribution/enso-engine-*/enso-*/bin/enso --run test/Table_Tests/ aggregates.and.distinct
[FAILED] [PostgreSQL]  Interactions Between various operations: [0/1, 327ms]

    - [FAILED] aggregates and distinct [380ms]
        Reason: An unexpected dataflow error ((Incomparable_Values.Error Nothing Nothing)) has been matched (at /home/devel/NetBeansProjects/enso/enso/test/Table_Tests/src/Common_Table_Operations/Integration_Tests.enso:70:13-72).
        at <enso> Incomparable_Values.type.handle_errors.handle.case handle(/home/devel/NetBeansProjects/enso/enso/built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Base/0.0.0-dev/src/Errors/Common.enso:278:47-101)
```
I can see the failure locally.

**Comment by JaroslavTulach** [comment]
> There is only a [single remaining failure](https://github.com/enso-org/enso/actions/runs/9869132872/job/27252332462?pr=10468#step:7:2416):

This failure has nothing to do with `Nothing` - it is just wrong error message - the culprit exception is from `ObjectComparator`:
```
          (v, u) -> {
            var result = compare_callback.execute(null, v, u);
            if (result.isNull()) {
              throw new CompareException(u, v);
            } else {
```
when comparing `BigInteger(13)` and `BigInteger(5)`. The following program now returns `[Nothing]` while previously it was returning `[-1]`:
```ruby
polyglot java import java.math.BigInteger

from Standard.Base import all
import Standard.Base.Internal.Ordering_Helpers.Default_Comparator

main =
  f = BigInteger.new "5"
  t = BigInteger.new "13"

  r = Default_Comparator.compare_callback f t
  [ r ]
```
Solved by a2c4def

**Comment by JaroslavTulach** [comment]
CI seems green, time for yet another round of benchmarks:
- [engine](https://github.com/enso-org/enso/actions/runs/9874878029)
- [stdlib](https://github.com/enso-org/enso/actions/runs/9874884130)

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering.enso [hunk 1]
```diff
@@ -25,20 +24,23 @@ from project.Data.Boolean import Boolean, False, True
    another `type` that satisfies the following definition:
 
    ```
-   type Comparator T
-       compare : T -> T -> (Ordering|Nothing)
+   type T_Comparator T
+       compare : T -> T -> (Ordering|Nothing ! Incomparable_Values)
```
Comment (Akirathan) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering.enso]: I think the correct type ascription is this one:
```suggestion
       compare : T -> T -> (Ordering|Nothing) ! Incomparable_Values
```

**Review by Akirathan — APPROVED** [review]
It is still not entirely clear to me why a custom comparator's `compare` method can throw `Incomparable_Values`. But I don't see any big problems with that. Looks good overall to me.

**Comment by JaroslavTulach** [comment]
> It is still not entirely clear to me why a custom comparator's `compare` method can throw `Incomparable_Values`. 

There are many places where comparator's `compare` delegates to `Ordering.compare`. `Ordering.compare` doesn't return `Nothing`, but yields an error. So either we relax the rules, or all the delegations will be complicated by recovering from the error.

>  I don't see any big problems with that.

Yes, relaxing the rules costs us nothing and the delegation is then very simple.

**Comment by JaroslavTulach** [comment]
> * [engine](https://github.com/enso-org/enso/actions/runs/9874878029)
> * [stdlib](https://github.com/enso-org/enso/actions/runs/9874884130)

Engine benchmarks are fine and my favorite stdlib benchmark looks way too optimistic:

too fast

One of the sorting benchmarks got speed up as well

sorting


Let's merge and see.

**Pull request merged**